### DISCUSSION TITLE
Proposal: Allow init/get to also have set and consider adding dispose for properties

### DISCUSSION THREAD

**Discussion by VWilcox2000** [dotnet/csharplang#6214] [Language Ideas]
c# 9.0 features new Init Only Setters that allow use of this.  This is a very nice option to shrink bloated constructors as you can now use 'this' in the property initialization, but it, of course, enforces set once only.  The ability to use 'this' allows very large classes, possibly split into multiple partial classes that need properties to have a reference back to the class to avoid huge constructors by using the init.  I would propose that allowing set in addition to init would provide the benefit of 'this' reference at initialization without the prevention of using the property as modifiable.  Also, in the cases of init's that do not use the value keyword, run them automatically during construction.  So if I have init { this._value = 6 } without a value, that is run on construction.  Alternatively, instead of 'init', 'default' keyword or similar could be used if it is felt that repurposing init without value specified is confusing.

Additionally, as the get/set domain is an easy place to add an optional keyword, a deinit or dispose keyword could be added which allows you to specify the cleanup code, if needed, for each property rather than having all properties referenced in a very complex Dispose method.

A typical class with 1 dependent property today:

      namespace CSharp;

      public sealed class SomeClass : IDisposable
      {
        public SomeClass()
        {
          this._item = new DependentClass(this);
        }

        private DependentClass _item;
        public DependentClass item
        {
          get => this._item;
          set
          {
            this._item = value;
          }
        }

        public void Dispose()
        {
          this.item.Dispose();
        }
      }

      public sealed class DependentClass : IDisposable
      {
        private SomeClass sc { get; }

        public DependentClass(SomeClass sc)
        {
          this.sc = sc;
        }

        public void Dispose()
        {
          // free up a theoretical system resource
        }
      }

Rather simple scenario, not hard to follow, but if 'SomeClass' is a large partial class with a number of dependent classes, the constructor can be very large -- and undesirably keep initialization code in the constructor rather than in the partial class file where it has effect.  If the compiler allows both init and set, we could replace it with:

      namespace CSharp;

      public sealed class SomeClass : IDisposable
      {
        public SomeClass()
        {
        }

        private DependentClass _item;
        public DependentClass item
        {
          init => new DependentClass(this);  // no use of 'value' so this is run on construction
          get => this._item;
          set => this._item = value; // item starts out set, but can be changed
        }

        public void Dispose()
        {
          this.item.Dispose();
        }
      }

      public sealed class DependentClass : IDisposable
      {
        private SomeClass sc { get; }

        public DependentClass(SomeClass sc)
        {
          this.sc = sc;
        }

        public void Dispose()
        {
          // free up a theoretical system resource
        }
      }

Another opportunity for quality of life improvement is to add deinit or dispose, I will use 'dispose':

      namespace CSharp;

      public sealed class SomeClass : IDisposable
      {
        public SomeClass()
        {
        }

        private DependentClass _item;
        public DependentClass item
        {
          init => new DependentClass(this);
          get => this._item;
          set => this._item = value;
          dispose => this._item.Dispose();
        }
      }

      public sealed class DependentClass : IDisposable
      {
        private SomeClass sc { get; }

        public DependentClass(SomeClass sc)
        {
          this.sc = sc;
        }

        public void Dispose()
        {
          // free up a theoretical system resource
        }
      }

For large, complex classes, this could shrink and help bring complex code together.  One caveat I see with the last is how to handle both an implicit and explicit IDisposable implementation.  I would think call both in a predefined order, likely the property disposes first.  I would bet the more typical case is just adding the dispose or two and adding the interface.

It would be nice to be able to only run init lazily at the first get -- this feature is potentially being provided for normal assignment with = in, I believe c# 11.  Perhaps init? => or lazy init =>  { ... } so the code is only run when they do their first get of the property's value.

Finally, I thought I had seen a potential feature to avoid using property backers in most cases, but I am having trouble finding it.  Perhaps I am confusing it with the option to use field so nameof is valid in future version.  If this is a duplicate, my apologies, but the property above can be simplified if the compiler keeps _item internally as the property's value as follows:

      namespace CSharp;

      public sealed class SomeClass : IDisposable
      {
        public SomeClass()
        {
        }

        public DependentClass item    // no backer, we are using innate value compiler would keep
        {
          init => new DependentClass(this);
          get;                                                       // automatic property -- return the value the compiler has
          set;                                                       // automatic -- set that value
          dispose => value.Dispose();
        }
      }

Where init assigned the value that the properties holds, get; just returns that value, set; sets it, and dispose takes a ref 'value' and does what it wants to do.  The get and set operators could also be overridden rather than automatic like so:

      namespace CSharp;

      public sealed class SomeClass : IDisposable
      {
        public SomeClass()
        {
        }

        public DependentClass item
        {
          init => new DependentClass(this);
          get
          {
            prop.value.UpdateLinks(this.someParameter);
            return prop.value;
          }
          set
          {
            prop.value = value;
            value.RealizeNewParent(this);
          }
          dispose => value.Dispose();
        }
      }

in this case choosing prop.value to assign or reference the property's value, but I would leave selection of such keywords to compiler / language experts.

**Comment by HaloFour** [comment]
`init` accessors already exist in C# and they mean something different.

**Reply by HaloFour** [reply]
As for lazy initialization: https://github.com/dotnet/csharplang/discussions/312

**Reply by VWilcox2000** [reply]
I believe in c# 11 they are implementing something to the effect of ?? = as well for lazy initialization for property values, but that 'lazy' keyword in #312 would work perfectly here in front of it.  I was suggesting if init did not use 'value', or was accompanied by 'set' it be run on construction, but perhaps 'default' would be a better keyword more clearly separated.  If you use Init with c# 9.0 implementation, you cannot also have a set -- that is the part I would like adjusted.  The main differentiator between init and set for inline-construction is the ability to use 'this' since:  public MyClass myClass { get; set; } = new MyClass(this);  is not possible.  It could be that c# 11.0's lazy instantiator will support it with public MyClass myClass { get; set; } ?? = new MyClass(this);  If so, that would achieve the same thing at least when lazy instantiation is desired.  If not, dual-purposing init: or adding something like default: would achieve it and also be OPTIONALLY lazy which would be important if the created property's class included say a background thread you needed to be running.

**Comment by VWilcox2000** [comment]
The current use of init may not be that different if you look at it this way:

      public sealed class A {
        private int _temp = 6;
        public int temp
        {
          get => this._temp;
          init
          {
            this._temp = value;
          }
        }
        ...

     This is allowing me from elsewhere to do:
        var a = new A
        {
          temp = 8
        };
        a.temp = 9 <-- but not this

Leaving that functionality alone, we can vary if if marked private or if we allow a set as below:

      public sealed class A {
        private int _temp = 6;
        public int temp
        {
          get => this._temp;
          init
          {
            this._temp = value ?? 32;
          }
          set
          {
            this._temp = value;
          }
        }
        ...

That would be redundant, but not if this slightly affects the behavior of init.  You can now give them the option of using the init on construction or not -- but if init is not performed, the init should be called with a null value so the user can set a default value.

        var a = new A
        {
          temp = 8
        };
        // a.temp == 8

        but if I do not:

        var a = new A();
        // a.temp == 32

This would not affect any existing code, but augments init.  If we further mark init as private, we can use it exclusively to set the value of the property as follows:

      public sealed class A {
        private int _temp;
        public int temp
        {
          get => this._temp;
          private init
          {
            this._temp = this.average + this.nextRandomValue;
          }
          set
          {
            this._temp = value;
          }
        }
        ...

Now, the use can no longer set property value on construction, but it is assigned a value in the init block.  Furthermore, this during assignment.  As mentioned previously by HaloFour based on proposal #312 , using the idea of adding lazy, you could make it private lazy init -- and it would be called on the first get or set.

      public sealed class A {
        private int _temp;
        public int temp
        {
          get => this._temp;
          private lazy init
          {
            this._temp = this.average + this.nextRandomValue + this.doSomeModeratelyLengthyCalculationsMaybe();
          }
          set
          {
            this._temp = value;
          }
        }
        ...

This would also lend itself well to making auto-properties more flexible:

      public sealed class A {
        // no backing field
        public int temp
        {
          private init
          {
            value = this.average + this.nextRandomValue;
          }
          get;
          set;
        }
        ...

which I would probably shorten to something like:

        public int temp { private init => this.average + this.nextRandomValue; get; set; }

or for readonly temp outside my class:

        public int temp { private init => this.average + this.nextRandomValue; get; private set; }

For lengthy record classes, I think a series of such auto property declarations would be easier to maintain and more straightforward to read.
Then, my final thought was the deinit or dispose function to the property to avoid having the clean-up code in the OnDispose but right with the property for example:

      public sealed class A {
        // no backing field
        public FileHandler dataFile
        {
          private init
          {
            value = FileHandler.Open(this.fileName);
          }
          get;
          // omitting set -- not desired
          dispose: value.Dispose();  // tell it to clean up those file handles
        }
        ...

        another syntactical option would be to make private init into:
        init => ... return the initial value to allow very brief code:

A compressed version of that could be something like:

        public FileHandler dataFile { private init => FileHandler.Open(this.fileName); get; dispose: dataFile.Dispose(); }

or, if we support dispose; as per get; set; doing the obvious, that could just be:

        public FileHandler dataFile { private init => FileHandler.Open(this.fileName); get; dispose; }

If properties use dispose: the language could either require that the class explicitly support IDisposable or automatically mark the class as IDisposable similar to how yield return implements IEnumerable/ator.

In a large class with multiple partial class source files, such brevity bringing the pertinent code for each property together into 1 or a few lines would be quite beneficial, I would think.

        Alternatively, init could be default: ..., but I think the scenarios above make init seem more intuitive to me at least.

**Reply by HaloFour** [reply]
`set` and `init` currently compile down to the same accessor method with slightly different metadata.  It's not possible to have both, not without changing the meaning of `init`, which would be a breaking change.  Similarly it's not possible to have the compiler call the `init` accessor with some default value if the consumer doesn't do so explicitly, as that would also be a breaking change.

**Reply by VWilcox2000** [reply]
Well, part of my point was, as they are not currently both supported simultaneously on a property, any current code would not feature them, hence not experiencing any breaks on previous code at the very least.  That said, if it was felt it would confuse coders after the change who might add both not realizing the implication, 'init' could be another appropriate keyword like default: or construct: which would add clarity at the expense of another keyword in the mix.

**Reply by PathogenDavid** [reply]
Portions of the .NET ecosystem (particularly the BCL) prioritize binary compatibility. Changing how `init` or `set` are exposed in compiled assemblies would cause issues in that regard.

Not an impossible issue to overcome, but it raises the bar for consideration and increases complexity.

**Reply by VWilcox2000** [reply]
Thanks for that explanation, @PathogenDavid.  Didn't think of using IL preview tool -- will keep that in mind in the future.

**Comment by zvrba** [comment]
Another use-case for allowing both `init` and `set`: I have a record holding columns of a query result fetched from the database. This record has init-only properties, which my ORM sets through reflection when reading data from the DB. Using records for this allows for caching and safely sharing instances of the record. I also get an equality check I can use before attempting to update the record in the DB. IOW, many goodies out of the box.

BUT. Some record types need editing. Blazor databinding requires that properties have setters. I attempted to do the following:

```
public record X {
    public int SomeProperty { get; init; protected set; }
}

public record X_EditModel : X {
    public new int SomeProperty { get => base.SomeProperty; set => base.SomeProperty = value; }
}
```

and stumbled upon the same problem. `set` and `init` cannot coexist. Current solution: copy-paste the record into `X_EditModel` and add lots of boiler-plate code for conversions between `X` and `X_EditModel`. IOW, very unsatisfactory solution.
